apps/bf_temp.c

```diff
@@ -0,0 +1,287 @@
+    methods_tempfile = NULL;
+}
+
+static int tempfile_new(BIO *b)
+{
+    TEMPFILE_CTX *temp;
+
+    temp = app_malloc(sizeof(*temp), "temp file");
+    temp->filename = NULL;
+    temp->tmp_filename = NULL;
+    BIO_set_data(b, temp);
+    BIO_set_init(b, 1);
+
+    return 1;
+}
+
+#ifdef _WIN32
+static int WIN32_rename(const char *from, const char *to);
+# define rename(from,to) WIN32_rename((from),(to))
+#endif
+
+static int tempfile_free(BIO *b)
+{
+    TEMPFILE_CTX *temp;
+
+    if (b == NULL)
+        return 0;
+
+    temp = BIO_get_data(b);
+    if (temp->filename != NULL && temp->tmp_filename != NULL) {
+#if !defined(__VMS)
+        rename(temp->tmp_filename, temp->filename);
+#endif
+    }
+    if (temp->filename)
+        OPENSSL_free(temp->filename);
+    if (temp->tmp_filename)
+        OPENSSL_free(temp->tmp_filename);
+    OPENSSL_free(temp);
+    BIO_set_data(b, NULL);
+    BIO_set_init(b, 0);
+
+    return 1;
+}
+
+static int tempfile_read(BIO *b, char *out, int outl)
+{
+    int ret = 0;
+    BIO *next = BIO_next(b);
+
+    if (out == NULL || outl == 0)
+        return 0;
+    if (next == NULL)
+        return 0;
+
+    ret = BIO_read(next, out, outl);
+    return ret;
+}
+
+static int tempfile_write(BIO *b, const char *in, int inl)
+{
+    BIO *next = BIO_next(b);
+
+    if ((in == NULL) || (inl <= 0))
+        return 0;
+    if (next == NULL)
+        return 0;
+
+    return BIO_write(next, in, inl);
+}
+
+#define RND_SIZE 16
+
+static void make_temp_filename(const char *orig, char *filename)
+{
+    int i, size = RND_SIZE;
+    char *p;
+    char set[] = "0123456789abcdefghijklmnopqrstuvwxyz"
+                 "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+
+    strcpy(filename, orig);
+    p = filename + strlen(orig);
+    *p++ = '.';
+
+    srand(time(NULL));
+    while (size--) {
+        i = rand() % (sizeof(set) - 1);
+        *p++ = set[i];
+    }
+
+    *p = '\0';
+
+    return;
+}
+
+static int istext(int format)
+{
+    return (format & B_FORMAT_TEXT) == B_FORMAT_TEXT;
+}
+
+static const char *modestr(char mode, int format)
+{
+    OPENSSL_assert(mode == 'a' || mode == 'r' || mode == 'w');
+
+    switch (mode) {
+    case 'a':
+        return istext(format) ? "a" : "ab";
+    case 'r':
+        return istext(format) ? "r" : "rb";
+    case 'w':
+        return istext(format) ? "w" : "wb";
+    }
+    /* The assert above should make sure we never reach this point */
+    return NULL;
+}
+
+static FILE *make_temp_file_ptr(const char *filename, int format, int *bflags,
+                                char **ret_tmp_filename)
+{
+    FILE *fp = NULL;
+    int fd = -1, mode, textmode;
+    char *tmp_filename = NULL;
+
+    /* create tmp file name */
+    tmp_filename = OPENSSL_malloc(strlen(filename) + RND_SIZE + 2);
+    if (tmp_filename == NULL)
+        goto err;
+    make_temp_filename(filename, tmp_filename);
+
+    /* create a tmp file sink BIO */
+    mode = O_WRONLY;
+#ifdef O_CREAT
+    mode |= O_CREAT;
+#endif
```

Comment:
You *must* add `O_EXCL` as well, to make sure you don't overwrite already existing files

Reply:
Does that `O_EXCL` available on all platforms?

Reply:
I'd also like to see `O_NOFOLLOW` conditionally added.

```
#ifdef O_EXCL
    mode |= O_EXCL
#endif 
#ifdef O_NOFOLLOW
    mode |= O_NOFOLLOW
#endif 
```

I've not seen a platform that hasn't defined `O_EXCL`, my experience is Unix biassed though.

Reply:
O_EXCL exists on every platform that has `open`